File: Multi-Plateform Interactive Hash Table Visualizer/Step 1/src/main/java/XHashtable.java
public class XHashtable {
    Node[] table;
    int size;
    public XHashtable(int size){
        this.size=size;
        table = new Node[size];
    }
    int hash(String key) {
        int hash = 0;
        for (int i = 0; i < key.length(); i++) {
            hash = (hash * 31 + key.charAt(i)) % size;
        }
        return hash;
    }
    public void add(String key) {
        int index = hash(key);
        // check if key already exists
        Node current = table[index];
        while (current != null) {
            if (current.key.equals(key)){
                return;
            }
            current = current.next;
        }
        Node newNode = new Node(key);
        newNode.next = table[index];
        table[index] = newNode;
    }
    public boolean delete(String key){
        int index = hash(key);
        Node current = table[index];
        Node prev = null;
        while (current != null){
            if (current.key.equals(key)){
                if (prev == null){
                    table[index] = current.next;
                }else{
                    prev.next = current.next;
                }
                return true;
            }
            prev = current;
            current = current.next;
        }
        return false;
    }
    public boolean find(String key){
        int index = hash(key);
        Node current = table[index];
        while (current != null){
            if (current.key.equals(key)){
                return true;
            }
            current = current.next;
        }
        return false;
    }
    public Node[] getTable() {
        return table;
    }
}

File: Multi-Plateform Interactive Hash Table Visualizer/Step 6/dart_flutter_app/lib/model/XHashtable.dart
import 'node.dart';

class XHashtable {
  final int size;
  late List<HashNode?> _table;
 
  XHashtable({this.size = 10}) {
    _table = List<HashNode?>.filled(size, null);
  }
  int hash(String key) {
    int hashValue = 0;
    for (int i = 0; i < key.length; i++) {
      hashValue = (hashValue * 31 + key.codeUnitAt(i)) % size;
    }
    return hashValue;
  }
  void add(String key) {
    final index = hash(key);
    HashNode? current = _table[index];
    while (current != null) {
      if (current.key == key) return;
      current = current.next;
    }
    final newNode = HashNode(key);
    newNode.next = _table[index];
    _table[index] = newNode;
  }
  bool delete(String key){
    final index = hash(key);
    HashNode? current =_table[index];
    HashNode? prev;
    while (current != null){
      if (current.key == key){
        if (prev == null){
          _table[index] = current.next;
        }else{
          prev.next = current.next;
        }
        return true;
      }
      prev = current;
      current = current.next;
    }
    return false;
  }
  bool find(String key){
        final index = hash(key);
        HashNode? current = _table[index];
        while (current != null){
            if (current.key == key){
                return true;
            }
            current = current.next;
        }
        return false;
    }
    List<HashNode?> getTable() => _table;
    List<List<String>?> getTableAsArray() {
      return _table.map((bucket) {
        if (bucket == null) return null;
        final result = <String>[];
        HashNode? current = bucket;
        while (current != null) {
          result.add(current.key);
          current = current.next;
        }
        return result;
      }).toList();
    }
}